### PR TITLE
Bug 1364894 - Upgrade Neutrino to v5

### PR DESCRIPTION
**Pull request by eliperelman**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2309)

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

File: neutrino-custom/index.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+const entries = ['index', 'admin', 'perf', 'logviewer', 'failureviewer', 'userguide'];
+
+module.exports = neutrino => {
+    const { source } = neutrino.options;
+    const { NODE_ENV, LOCAL_WATCH } = process.env;
+
+    neutrino.config
+        // Environment-specific configuration
+        .when(NODE_ENV === 'development' && !LOCAL_WATCH, () => neutrino.use(development))
+        .when(NODE_ENV === 'production' || LOCAL_WATCH || NODE_ENV === 'test', () => neutrino.use(production))
```
Comment (edmorley) [inline neutrino-custom/index.js]: In the ui-tests Travis run chunk, we do both a `yarn test` and a `yarn run build`. From reading this it seems the test run is already very similar to the prod build. Do we still need to run both commands in the Travis run, or can we skip the `yarn run build` to save CI end to end time? Curious what is done in other projects..?

Also, it seems like we have a lot of permutations here. I have a PR locally that gets full-stack auth working with the standard development config, making local-watch.js redundant. Do you think we can get rid of test vs non-test too? It seems like there are really three configs: 1) dev, 2) test, 3) prod -- where test should either follow dev or prod rather than be a subset of both?

File: neutrino-custom/index.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        .when(NODE_ENV === 'test' && !LOCAL_WATCH, () => neutrino.use(test), () => neutrino.use(notTest))
+        .when(LOCAL_WATCH, () => neutrino.use(localWatch))
+
+        // Common configuration
+        .plugins
+            .delete('html').end()
+        .plugin('provide')
+            // Now provide a few globals (providing angular here does not seem to work, though)
```
Comment (edmorley) [inline neutrino-custom/index.js]: I'm presuming at some point these should be converted to narrower imports? If so, could you add a comment to that effect?

File: neutrino-custom/index.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                treeherder: require.resolve(join(source, 'js/treeherder.js')),
+                treeherderApp: require.resolve(join(source, 'js/treeherder_app.js')),
+                perf: require.resolve(join(source, 'js/perf.js')),
+                admin: require.resolve(join(source, 'js/admin.js')),
+                failureViewerApp: require.resolve(join(source, 'js/failureviewer.js')),
+                logViewerApp: require.resolve(join(source, 'js/logviewer.js')),
+                userguideApp: require.resolve(join(source, 'js/userguide.js'))
+            }]).end()
+        .entry('index')
+            .delete(require.resolve('babel-polyfill')).end()
```
Comment (edmorley) [inline neutrino-custom/index.js]: Why is this needed/what does it do? Could you add a comment?

File: neutrino-custom/index.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    neutrino.config
+        // Environment-specific configuration
+        .when(NODE_ENV === 'development' && !LOCAL_WATCH, () => neutrino.use(development))
+        .when(NODE_ENV === 'production' || LOCAL_WATCH || NODE_ENV === 'test', () => neutrino.use(production))
+        .when(NODE_ENV === 'test' && !LOCAL_WATCH, () => neutrino.use(test), () => neutrino.use(notTest))
+        .when(LOCAL_WATCH, () => neutrino.use(localWatch))
+
+        // Common configuration
+        .plugins
+            .delete('html').end()
```
Comment (edmorley) [inline neutrino-custom/index.js]: I'm presuming this is due to the custom html loader below for the angular partials thing? If so, could you add a comment?

File: neutrino-custom/not-test.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+            'angular-route', 'angular-sanitize', 'angular-toarrayfilter', 'angular-ui-bootstrap',
+            'angular-ui-router', 'bootstrap/dist/js/bootstrap', 'hawk', 'jquery', 'jquery.scrollto',
+            'js-yaml', 'mousetrap', 'react', 'react-dom', 'taskcluster-client'
+        ])
+        .end()
+    // Ensure chunkhashes for unchanged chunks don't change -- this allows the bundled vendor
+    // file to keep its hash and stay cached in the user's browser when it hasn't been updated.
+    .output
+        .chunkFilename('[chunkhash].[id].chunk.js').end()
+    .plugin('hash')
```
Comment (edmorley) [inline neutrino-custom/not-test.js]: I'm presuming this will be made redundant if https://github.com/mozilla-neutrino/neutrino-dev/issues/99 is implemented? If so, could you add a comment linking to it?

File: neutrino-custom/production.js [hunk 6]
```diff
@@ -1,48 +1,26 @@
...
-        });
-
-    // Likewise for this clean plugin:
-    neutrino.config.plugin('clean')
-        .use(CleanPlugin, [DIST], { root: CWD } );
-
+    neutrino.use(copy, {
+        patterns: [
+            { context: neutrino.options.source, from: '**' },
+            { from: './contribute.json', to: 'contribute.json' }
```
Comment (edmorley) [inline neutrino-custom/production.js]: I'm going to move `contribute.json` into the `ui/` directory in another PR so we don't have to special-case this.

File: neutrino-custom/index.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                .use('raw')
+                    .loader(require.resolve('raw-loader')).end()
+        // The page templates should be excluded from the rule
+        // so that they don't collide with the html plugins:
+        .exclude
+            .merge(entries.map(entry => join(source, `${entry}.html`))).end()
+            .uses.delete('file');
+
+    // Add several other treeherder entry points:
+    entries.forEach(entry => neutrino.config
```
Comment (edmorley) [inline neutrino-custom/index.js]: Could this be replaced by `neutrino-middleware-html-template`?

File: package.json [hunk 8]
```diff
@@ -38,25 +39,50 @@
...
+    "neutrino-middleware-eslint": "^5.2.0",
+    "neutrino-preset-karma": "^5.2.0",
+    "neutrino-preset-react": "^5.2.0",
     "ngreact": "0.3.0",
     "raw-loader": "0.5.1",
     "react": "15.3.1",
     "react-addons-test-utils": "15.3.1",
     "react-dom": "15.3.1",
     "taskcluster-client": "1.6.3",
+    "webpack": "^2.3.2",
```
Comment (edmorley) [inline package.json]: webpack will be pulled in via the neutrino dependencies. What are your thoughts about leaving neutrino to set the version instead? The versions will still be pinned via `yarn.lock`.

File: package.json [hunk 9]
```diff
@@ -38,25 +39,50 @@
...
   "scripts": {
-    "build": "./node_modules/neutrino/bin/neutrino build --presets ./neutrino-custom/production.js",
-    "lint": "./node_modules/eslint/bin/eslint.js ui/",
-    "start": "./node_modules/neutrino/bin/neutrino start --presets ./neutrino-custom/development.js",
-    "start:local": "./node_modules/neutrino/bin/neutrino start --presets ./neutrino-custom/local-watch.js",
-    "start:stage": "SERVICE_DOMAIN=https://treeherder.allizom.org ./node_modules/neutrino/bin/neutrino start --presets ./neutrino-custom/development.js",
-    "test": "./node_modules/neutrino/bin/neutrino test --presets ./neutrino-custom/test.js",
-    "test:watch": "./node_modules/neutrino/bin/neutrino test --watch --presets ./neutrino-custom/test.js"
+    "build": "./node_modules/neutrino/bin/neutrino build",
+    "lint": "NODE_ENV=development ./node_modules/eslint/bin/eslint.js --ext .js,.jsx ui/",
```
Comment (edmorley) [inline package.json]: Can `--ext .js,.jsx` be put in the config file instead, or does eslint not support that?

Also, at the moment we have the eslint version set in two places - the top level package.json entry, and via the neutrino dependency. It seems unfortunate to have to keep those in sync. But if we remove the top-level eslint entry, (a) we get a peer dependency warning (or at least we did with neutrino v4), (b) due to yarnpkg/yarn#2874 the bin script won't be created. Is there a better way to do this? Could neutrino add eslint to it's own package.json `bin` section per https://github.com/yarnpkg/yarn/issues/2874#issuecomment-290668434 ?

File: ui/js/reactrevisions.jsx [hunk 10]
```diff
@@ -4,9 +4,9 @@ const MoreRevisionsLink = (props) => (
     <li>
         <a href={props.href}
            data-ignore-job-clear-on-click={true}
-           target='_blank'>
```
Comment (edmorley) [inline ui/js/reactrevisions.jsx]: I presume these are required since previously we were accidentally not including .jsx files in the eslint run?

File: neutrino-custom/lint.js [hunk 11]
```diff
@@ -1,92 +1,77 @@
...
+                es6: true,
+                jsx: true,
+                impliedStrict: false
             }
-        }));
-};
-
+        },
+        extends: 'eslint:recommended',
+        rules: {
```
Comment (edmorley) [inline neutrino-custom/lint.js]: I'm looking forward to using a standard style at some point in the future (eg the airbnb one) so we can then remove this :-)

File: neutrino-custom/production.js [hunk 12]
```diff
@@ -1,48 +1,26 @@
...
-    // Define the service domain globally so that window.thServiceDomain can be set:
-    neutrino.config.plugin('define')
-        .use(webpack.DefinePlugin, {
-            SERVICE_DOMAIN: JSON.stringify(SERVICE_DOMAIN)
-        });
-
+    neutrino.config
+        .plugin('minify')
+        // prevents some minification errors
+            .tap(() => [{ evaluate: false }]).end()
```
Comment (edmorley) [inline neutrino-custom/production.js]: Do you know why we need this?

File: package.json [hunk 13]
```diff
@@ -38,25 +39,50 @@
...
+    "start": "./node_modules/neutrino/bin/neutrino start",
+    "start:local": "LOCAL_WATCH=true ./node_modules/neutrino/bin/neutrino start",
```
Comment (edmorley) [inline package.json]: Setting environment variables like this doesn't work on Windows. (`start:stage` is already broken there, but now `start:local` and `lint` are too)

The alternatives seem to be to use [cross-env](https://www.npmjs.com/package/cross-env) or else switch to flag options (eg `neutrino start -- --watch-local`) instead? Thoughts?

File: neutrino-custom/index.js (hunk 1 above)
Comment (eliperelman) [inline neutrino-custom/index.js]: To answer the first question, yes, I would skip the creating a build and just do the test build (and a lint run).

To answer your second question, I totally agree. I tried to make the configurations converge as much as possible here without doing anything breaking (mostly upgrade now, change later).

File: neutrino-custom/index.js (hunk 4 above)
Comment (eliperelman) [inline neutrino-custom/index.js]: This was already being done prior for the partials, but the comment for it is down later. I'll move this chain to the relevant comment area.

Comment (eliperelman) [inline neutrino-custom/index.js]: This was already being done prior for the partials, but the comment for it is down later. I'll move this chain to the relevant comment area.

File: neutrino-custom/index.js (hunk 2 above)
Comment (eliperelman) [inline neutrino-custom/index.js]: At some point you should move to modules instead of scripts, which will negate the need for these providers. But I can comment accordingly.

File: neutrino-custom/index.js (hunk 3 above)
Comment (eliperelman) [inline neutrino-custom/index.js]: Neutrino v5 added babel-polyfill by default to the index entry point. Treeherder doesn't use this right now, so I removed it to avoid breaking anything on that entry. I will comment.

File: neutrino-custom/index.js (hunk 7 above)
Comment (eliperelman) [inline neutrino-custom/index.js]: It's possible, but I wanted to avoid making any major breakages with the way files were being generated.

File: neutrino-custom/not-test.js (hunk 5 above)
Comment (eliperelman) [inline neutrino-custom/not-test.js]: Correct. Will do.

File: package.json (hunk 8 above)
Comment (edmorley) [inline package.json]: Yeah I couldn't decide which was better. Being explicit about deps seems preferred, though potentially having two different versions of webpack in use seems bad too (if the versions are accidentally not kept in sync). I'm happy to defer to whatever you think is best.

File: package.json (hunk 9 above)
Comment (eliperelman) [inline package.json]: Regarding `ext`, right now they must be passed on the command line, and do not have the ability to be set in the config file.

I agree with the eslint dep; I would remove from your package.json. Adding it to the bin is an interesting idea, and hadn't considered it before, and I think it would be a positive change. It wouldn't be a part of neutrino, but rather defined by neutrino-middleware-eslint. I'll file an issue for that.

File: ui/js/reactrevisions.jsx (hunk 10 above)
Comment (eliperelman) [inline ui/js/reactrevisions.jsx]: It's not required, but double quotes are considered idiomatic for HTML/JSX. This comes along with the React preset.

File: neutrino-custom/lint.js (hunk 11 above)
Comment (eliperelman) [inline neutrino-custom/lint.js]: Agreed. I need to revise our shared configuration.

File: neutrino-custom/production.js (hunk 12 above)
Comment (eliperelman) [inline neutrino-custom/production.js]: I'm not sure, this was added by @caseywilliams. Casey?

File: package.json (hunk 13 above)
Comment (eliperelman) [inline package.json]: Rawr, I always forget about Windows. I like cross-env; arg sniffing would work also, but we are already using env vars, so I'd probably go that route.

File: package.json (hunk 8 above)
Comment (eliperelman) [inline package.json]: I usually err on the side of explicit over implicit. Relying on a transitive dependency has caused some painful errors in the past. I like to declare what I actually depend on, and this is a case where you depend on Neutrino **and** Webpack. Might as well be up-front about it, imo.

File: package.json (hunk 9 above)
Comment (edmorley) [inline package.json]: > Regarding ext, right now they must be passed on the command line, and do not have the ability to be set in the config file.

I was going to file an eslint issue for adding an .eslintrc option, but looks like there is a reason why it's not possible:
https://github.com/eslint/eslint/issues/3469

That said, perhaps they could default to including .jsx, to avoid the need to override. I've filed:
https://github.com/eslint/eslint/issues/8399

File: neutrino-custom/production.js (hunk 12 above)
Comment (caseywilliams) [inline neutrino-custom/production.js]: Without this, I got some errors in minified code that were just too cryptic for me to figure out - for example: 

```
TypeError: Cannot read property 'revision' of undefined
    at index.7905b7e….bundle.js:35
    at Al (vendor.103bf08….bundle.js:35)
    at vendor.103bf08….bundle.js:35
    at Rl.$eval (vendor.103bf08….bundle.js:35)
    at Rl.$digest (vendor.103bf08….bundle.js:35)
    at Rl.$apply (vendor.103bf08….bundle.js:35)
    at vendor.103bf08….bundle.js:35
    at Al (vendor.103bf08….bundle.js:35)
    at vendor.103bf08….bundle.js:35
```

After some guesssing and checking, I determined this was coming from the eval option (corresponds to this [package](https://github.com/babel/babili/tree/master/packages/babel-plugin-minify-constant-folding)). I'm thinking this may be easier to fix once sourcemaps are present, but if anyone has an idea what could be making this happen, I'm happy to make a PR with a fix!

Comment (eliperelman) [inline neutrino-custom/production.js]: Neutrino v5 switches the devtool in development from `eval` to `source-map`. Maybe this isn't needed any more with v5?

File: package.json (hunk 13 above)
Comment (edmorley) [inline package.json]: I've just seen that the cross-env author is now recommending people use the new webpack 2 `--env` options instead, so I've filed:
https://github.com/mozilla-neutrino/neutrino-dev/issues/165

File: neutrino-custom/lint.js [hunk 14]
```diff
@@ -1,92 +1,77 @@
 'use strict';
+const eslint = require('neutrino-middleware-eslint');
 
-const merge = require('deepmerge');
```
Comment (edmorley) [inline neutrino-custom/lint.js]: deepmerge is now unused be Treeherder itself, so can be removed from package.json

File: neutrino-custom/production.js (hunk 6 above)
Comment (edmorley) [inline neutrino-custom/production.js]: I've done this in #2321. 
Once that lands, does it mean we can use stock Neutrino behaviour here, and remove this step?

File: ui/js/reactrevisions.jsx (hunk 10 above)
Comment (edmorley) [inline ui/js/reactrevisions.jsx]: Ah I presume we were inadvertently clobbering the react preset previously.

File: package.json (hunk 9 above)
Comment (edmorley) [inline package.json]: Re the `NODE_ENV=development` addition here, it would be great if Neutrino did this for us. I've filed mozilla-neutrino/neutrino-dev#192.

File: .eslintrc.js [hunk 15]
```diff
@@ -1,5 +1,7 @@
-const path = require('path');
-const Neutrino = require('neutrino');
-const api = new Neutrino([path.resolve('./neutrino-custom/development.js')]);
+const { Neutrino } = require('neutrino');
+const { neutrino } = require('./package.json');
 
-module.exports = api.custom.eslintrc();
+const api = Neutrino(neutrino.options);
+
+neutrino.use.map(middleware => api.use(require(middleware)));
```
Comment (edmorley) [inline .eslintrc.js]: This differs slightly from the `.eslintrc.js` example given in the Neutrino docs. If this is preferred over the form given there, could you update the docs?
https://neutrino.js.org/middleware/neutrino-middleware-eslint/#eslintrc-config

File: package.json [hunk 16]
```diff
@@ -38,25 +39,50 @@
...
     "react": "15.3.1",
     "react-addons-test-utils": "15.3.1",
     "react-dom": "15.3.1",
     "taskcluster-client": "1.6.3",
+    "webpack": "^2.3.2",
     "webpack-md5-hash": "0.0.5"
   },
+  "neutrino": {
+    "use": [
+      "./neutrino-custom/lint.js",
```
Comment (edmorley) [inline package.json]: Why does the custom lint entry go before the neutrino react preset?
I would have thought it would need to go after, it it were overriding it?

File: package.json [hunk 17]
```diff
@@ -38,25 +39,50 @@
...
-    "start:stage": "SERVICE_DOMAIN=https://treeherder.allizom.org ./node_modules/neutrino/bin/neutrino start --presets ./neutrino-custom/development.js",
-    "test": "./node_modules/neutrino/bin/neutrino test --presets ./neutrino-custom/test.js",
-    "test:watch": "./node_modules/neutrino/bin/neutrino test --watch --presets ./neutrino-custom/test.js"
+    "build": "./node_modules/neutrino/bin/neutrino build",
+    "lint": "NODE_ENV=development ./node_modules/eslint/bin/eslint.js --ext .js,.jsx ui/",
+    "start": "./node_modules/neutrino/bin/neutrino start",
+    "start:local": "LOCAL_WATCH=true ./node_modules/neutrino/bin/neutrino start",
+    "start:stage": "SERVICE_DOMAIN=https://treeherder.allizom.org ./node_modules/neutrino/bin/neutrino start",
+    "test": "./node_modules/neutrino/bin/neutrino test",
+    "test:watch": "./node_modules/neutrino/bin/neutrino test --watch --use"
```
Comment (edmorley) [inline package.json]: Why the empty `--use`? I was presuming this was to reset the list given in package.json, but it's not done for the plain (non-watch) test command, so I'm not sure why so for this one?

File: neutrino-custom/not-test.js (hunk 5 above)
Comment (edmorley) [inline neutrino-custom/not-test.js]: Also, I know this is pre-existing, but I don't think the hashing should be done in development, according to https://webpack.js.org/guides/caching/ 

> Don’t use [chunkhash] in development since this will increase compilation time. Separate development and production configs and use [name].js for development and [name].[chunkhash].js in production. 

ie: not-test.js probably shouldn't exist, and all of this be moved into production.js perhaps? (Though perhaps the vendor splitting should be in the common config?)
